src/viewport_renderer.rs

```diff
@@ -40,23 +44,110 @@ pub struct Vertex {
 // FIXME: @Optimization Determine u16/u32 dynamically per geometry to
 // save memory
 /// The geometry indices as uploaded on the GPU.
-pub type Index = u32;
+pub type RendererIndex = u32;
 
-/// The geometry containing index and vertex data in the format as
+/// The geometry containing index and vertex data in flat format as
 /// will be uploaded on the GPU.
 #[derive(Debug, Clone, PartialEq)]
-pub struct Geometry {
-    indices: Option<Vec<Index>>,
-    vertex_data: Vec<Vertex>,
+pub struct RendererGeometry {
+    indices: Option<Vec<RendererIndex>>,
+    vertex_data: Vec<RendererVertex>,
 }
 
-impl Geometry {
+impl RendererGeometry {
+    /// Construct flat geometry with same-length per-vertex data from
+    /// variable length data `Geometry`
+    pub fn from_geometry(geometry: &Geometry) -> Self {
+        let vertices = geometry.vertices();
+        if let Some(normals) = geometry.normals() {
+            // Duplicate either vertices or normals, whichever list
+            // was shorter
+            let vertices_len = vertices.len();
+            let normals_len = normals.len();
+            let len = vertices_len.max(normals_len);
+
+            let indices_len = geometry.triangle_face_indices_len();
+            let mut indices: Vec<RendererIndex> = Vec::with_capacity(indices_len);
+            let mut vertex_positions: Vec<Point3<f32>> = vec![Point3::origin(); len];
+            let mut vertex_normals: Vec<Vector3<f32>> = vec![Vector3::zeros(); len];
+
+            match vertices_len.cmp(&normals_len) {
+                Ordering::Less => {
```

Comment:
I should test this with a compressed version of `geometry::uv_cube()` which has more normals than vertices. The conversion should decompress the vertices correctly.